### ISSUE TITLE
Metal: --embeddings with pooling still builds and computes the LM head, producing a multi-GiB discarded tensor and returning all-NaN embeddings on long inputs

### ISSUE THREAD

**Issue by ZohnNasse** [ggml-org/llama.cpp#27784]
<h2 dir="ltr">Summary</h2>

<p dir="ltr">When <code>llama-server --embeddings</code> runs a pooled embedding model on the Metal backend, the graph still contains <code>MUL_MAT result_output</code> (the LM head). Its output is never read — <code>result_embd_pooled</code> is a <code>GET_ROWS</code> from <code>result_norm</code> — but it is fully materialized on every request.</p>

<p dir="ltr">For <code>Qwen3-Embedding-0.6B</code> (vocab 151669) at 7343 tokens that discarded tensor is f32 <code>[151669, 7343]</code> = 4,454,821,868 bytes (4.15 GiB), and it grows linearly with input length. Past some input size the Metal backend then produces out-of-bounds behavior and the server returns all-NaN vectors. Gating the LM head behind an <code>embeddings + pooling</code> check eliminates the failure and drops the MTL0 compute buffer by roughly 10×.</p>

<h2 dir="ltr">Symptom</h2>

<p dir="ltr">Embeddings come back as all-NaN. <code>nlohmann/json</code> serializes NaN as <code>null</code>, so a client sees a vector of <code>null</code> rather than an error.</p>

<ul dir="ltr">
<li><code>/health</code> continues returning 200</li>
<li>no Metal error is logged</li>
<li>no recovery short of restarting the process</li>
<li><code>-ngl 0</code> (CPU) is completely clean</li>
</ul>

<p dir="ltr">The onset depends on how far past the threshold the input is:</p>

<ul dir="ltr">
<li><strong>Near the threshold</strong> (~7343 tokens at vocab 151669): the triggering request itself returns a <strong>valid</strong> vector, and every request after it — including a 4-token input — returns all-NaN for the lifetime of the process. This is consistent with the overrun landing in a mapped neighbor allocation rather than in the triggering request's own output.</li>
<li><strong>Well past it</strong> (28,001 tokens): the triggering request is already all-NaN itself.</li>
</ul>

<p dir="ltr">Worth stating explicitly because the near-threshold case is the dangerous one — the request that breaks the server looks successful, so a pipeline that validates only the newest response never sees a failure.</p>

<h2 dir="ltr">Environment</h2>

<p dir="ltr">Reproduced on unmodified upstream master:</p>

<ul dir="ltr">
<li><code>build 1 (192067b)</code>, AppleClang 21.0.0.21000101, Darwin arm64, shallow clone of <code>ggml-org/llama.cpp</code>, <code>-DGGML_METAL=ON -DCMAKE_BUILD_TYPE=Release</code></li>
<li>Apple M5 Max, 128 GB unified memory (110100 MiB visible to Metal), macOS</li>
<li><code>llama-server --embeddings</code>, <code>Qwen3-Embedding-0.6B</code> f16, <code>-c 32768 -np 1 -b 8192 -ub 8192 -ngl 99</code></li>
</ul>

<p dir="ltr">Originally found on two independent forks; the pristine build above was made specifically to rule out fork-local Metal changes, and reproduces identically.</p>

<h2 dir="ltr">Reproduction</h2>

<div role="group" aria-label="bash 코드" tabindex="0"><div><div></div></div><div>bash</div><div><pre style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono);"><code class="language-bash" style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono); white-space: pre;"><span style="color: rgb(0, 81, 194);">llama-server</span> <span style="color: rgb(0, 128, 0);">-m</span> <span style="color: rgb(0, 128, 0);">Qwen3-Embedding-0.6B-f16.gguf</span> <span style="color: rgb(0, 128, 0);">--embeddings</span> <span style="color: rgb(0, 128, 128);">\</span>
  <span style="color: rgb(0, 128, 0);">-ngl</span> <span style="color: rgb(0, 128, 128);">99</span> <span style="color: rgb(0, 128, 0);">-c</span> <span style="color: rgb(0, 128, 128);">32768</span> <span style="color: rgb(0, 128, 0);">-b</span> <span style="color: rgb(0, 128, 128);">8192</span> <span style="color: rgb(0, 128, 0);">-ub</span> <span style="color: rgb(0, 128, 128);">8192</span> <span style="color: rgb(0, 128, 0);">-np</span> <span style="color: rgb(0, 128, 128);">1</span> <span style="color: rgb(0, 128, 0);">--host</span> <span style="color: rgb(0, 128, 128);">127.0.0.1</span> <span style="color: rgb(0, 128, 0);">--port</span> <span style="color: rgb(0, 128, 128);">8084</span>

<span style="color: rgb(0, 81, 194);">python3</span> <span style="color: rgb(0, 128, 0);">-c</span> <span style="color: rgb(0, 128, 0);">"print('가나다라마바사한글테스트문장입니다'*2000, end='')"</span> &gt; <span style="color: rgb(0, 128, 0);">/tmp/long.txt</span>

<span style="color: rgb(0, 81, 194);">curl</span> <span style="color: rgb(0, 128, 0);">-s</span> <span style="color: rgb(0, 128, 0);">localhost:8084/v1/embeddings</span> <span style="color: rgb(0, 128, 0);">-H</span> <span style="color: rgb(0, 128, 0);">'Content-Type: application/json'</span> <span style="color: rgb(0, 128, 128);">\</span>
  <span style="color: rgb(0, 128, 0);">-d</span> <span style="color: rgb(0, 128, 0);">"</span><span style="color: rgb(43, 48, 59);">$(</span><span style="color: rgb(0, 81, 194);">python3</span><span style="color: rgb(0, 128, 0);"> -c "import json;print(json.dumps({'input':open('/tmp/long.txt').read(),'model':'emb'}))"</span><span style="color: rgb(43, 48, 59);">)</span><span style="color: rgb(0, 128, 0);">"</span> <span style="color: rgb(0, 128, 128);">\</span>
  | <span style="color: rgb(0, 81, 194);">python3</span> <span style="color: rgb(0, 128, 0);">-c</span> <span style="color: rgb(0, 128, 0);">"import sys,json;d=json.load(sys.stdin);v=d['data'][0]['embedding'];print('prompt_tokens',d['usage']['prompt_tokens'],'nan',sum(1 for x in v if x is None or x!=x),'/',len(v))"</span>

<span style="color: rgb(0, 81, 194);">curl</span> <span style="color: rgb(0, 128, 0);">-s</span> <span style="color: rgb(0, 128, 0);">localhost:8084/v1/embeddings</span> <span style="color: rgb(0, 128, 0);">-H</span> <span style="color: rgb(0, 128, 0);">'Content-Type: application/json'</span> <span style="color: rgb(0, 128, 128);">\</span>
  <span style="color: rgb(0, 128, 0);">-d</span> <span style="color: rgb(0, 128, 0);">'{"input":"hello","model":"emb"}'</span> <span style="color: rgb(0, 128, 128);">\</span>
  | <span style="color: rgb(0, 81, 194);">python3</span> <span style="color: rgb(0, 128, 0);">-c</span> <span style="color: rgb(0, 128, 0);">"import sys,json;v=json.load(sys.stdin)['data'][0]['embedding'];print('nan',sum(1 for x in v if x is None or x!=x),'/',len(v))"</span></code></pre></div></div>

<p dir="ltr">Observed on the pristine build:</p>

<div role="group" aria-label="코드" tabindex="0"><div><div></div></div><div><pre style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono);"><code style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono); white-space: pre-wrap;">prompt_tokens 28001 nan 1024 / 1024
nan 1024 / 1024</code></pre></div></div>

<p dir="ltr">Any non-Latin text works; the point is only to exceed the token threshold.</p>

<p dir="ltr">Adding <code>--no-kv-offload</code> converts the silent corruption into a deterministic hard fault:</p>

<div role="group" aria-label="코드" tabindex="0"><div><div></div></div><div><pre style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono);"><code style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono); white-space: pre-wrap;">ggml_metal_synchronize: command buffer 1 failed with status 5
error: Caused GPU Address Fault Error (kIOGPUCommandBufferCallbackErrorPageFault)</code></pre></div></div>

<p dir="ltr">which is what made the offending node identifiable. The reason it does not normally fault is that the overrun lands inside a mapped neighbor allocation — the KV cache. Pull KV off the GPU and that region is unmapped, so the same access page-faults.</p>

<h2 dir="ltr">Locating the node</h2>

<p dir="ltr">With <code>-nkvo</code> for determinism and <code>GGML_METAL_GRAPH_DEBUG=2</code> for per-node logging (note: this requires <code>-lv 5</code> — <code>LOG_LEVEL_DEBUG</code> is 5 in <code>common/log.h</code> and <code>common/log.cpp</code> drops all DEBUG output when verbosity &lt; 5, so <code>-lv 4</code> silently logs nothing):</p>

<div role="group" aria-label="코드" tabindex="0"><div><div></div></div><div><pre style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono);"><code style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono); white-space: pre-wrap;">node[  14] - MUL_MAT      result_output
  src0 = [ 1024, 151669, 1, 1] f16
  src1 = [ 1024,   7343, 1, 1] f32
  node = [151669,  7343, 1, 1] f32   nb = [4, 606676, 4454821868, ...]
node[  15] - GET_ROWS     result_embd_pooled
  src0 = result_norm
E ggml_metal_synchronize: command buffer 1 failed with status 5</code></pre></div></div>

<p dir="ltr">Node 15 confirms the pooled embedding comes from <code>result_norm</code>, so node 14's 4.15 GiB output is dead weight.</p>

<h2 dir="ltr">Cause</h2>

<p dir="ltr"><code>src/models/qwen3.cpp</code> appends the LM head with no check on <code>cparams.embeddings</code>:</p>

<div role="group" aria-label="cpp 코드" tabindex="0"><div><div></div></div><div>cpp</div><div><pre style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono);"><code class="language-cpp" style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono); white-space: pre;">cur = <span style="color: rgb(0, 81, 194);">build_lora_mm</span><span style="color: rgb(43, 48, 59);">(</span>model<span style="color: rgb(43, 48, 59);">.</span>output<span style="color: rgb(43, 48, 59);">,</span> cur<span style="color: rgb(43, 48, 59);">,</span> model<span style="color: rgb(43, 48, 59);">.</span>output_s<span style="color: rgb(43, 48, 59);">);</span>
<span style="color: rgb(0, 81, 194);">cb</span><span style="color: rgb(43, 48, 59);">(</span>cur<span style="color: rgb(43, 48, 59);">,</span> <span style="color: rgb(0, 128, 0);">"result_output"</span><span style="color: rgb(43, 48, 59);">,</span> -<span style="color: rgb(0, 128, 128);">1</span><span style="color: rgb(43, 48, 59);">);</span>
res<span style="color: rgb(43, 48, 59);">-&gt;</span>t_logits = cur<span style="color: rgb(43, 48, 59);">;</span>
<span style="color: rgb(0, 81, 194);">ggml_build_forward_expand</span><span style="color: rgb(43, 48, 59);">(</span>gf<span style="color: rgb(43, 48, 59);">,</span> cur<span style="color: rgb(43, 48, 59);">);</span></code></pre></div></div>

<p dir="ltr"><code>set_outputs</code> in <code>src/llama-graph.cpp</code> likewise marks <code>t_logits</code> as a graph output unconditionally.</p>

<h2 dir="ltr">Proposed fix (local, per-architecture)</h2>

<div role="group" aria-label="cpp 코드" tabindex="0"><div><div></div></div><div>cpp</div><div><pre style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono);"><code class="language-cpp" style="color: rgb(20, 24, 31); background: transparent; font-family: var(--font-mono); white-space: pre;"><span style="color: rgb(110, 118, 135);">// logits are unused in embeddings mode with pooling</span>
<span style="color: rgb(129, 0, 194);">if</span> <span style="color: rgb(43, 48, 59);">(</span>!cparams<span style="color: rgb(43, 48, 59);">.</span>embeddings || pooling_type == LLAMA_POOLING_TYPE_NONE<span style="color: rgb(43, 48, 59);">)</span> <span style="color: rgb(43, 48, 59);">{</span>
    cur = <span style="color: rgb(0, 81, 194);">build_lora_mm</span><span style="color: rgb(43, 48, 59);">(</span>model<span style="color: rgb(43, 48, 59);">.</span>output<span style="color: rgb(43, 48, 59);">,</span> cur<span style="color: rgb(43, 48, 59);">,</span> model<span style="color: rgb(43, 48, 59);">.</span>output_s<span style="color: rgb(43, 48, 59);">);</span>
    <span style="color: rgb(0, 81, 194);">cb</span><span style="color: rgb(43, 48, 59);">(</span>cur<span style="color: rgb(43, 48, 59);">,</span> <span style="color: rgb(0, 128, 0);">"result_output"</span><span style="color: rgb(43, 48, 59);">,</span> -<span style="color: rgb(0, 128, 128);">1</span><span style="color: rgb(43, 48, 59);">);</span>
    res<span style="color: rgb(43, 48, 59);">-&gt;</span>t_logits = cur<span style="color: rgb(43, 48, 59);">;</span>
    <span style="color: rgb(0, 81, 194);">ggml_build_forward_expand</span><span style="color: rgb(43, 48, 59);">(</span>gf<span style="color: rgb(43, 48, 59);">,</span> cur<span style="color: rgb(43, 48, 59);">);</span>
<span style="color: rgb(43, 48, 59);">}</span></code></pre></div></div>

<p dir="ltr">Both <code>cparams</code> and <code>pooling_type</code> are already members of <code>llm_graph_context</code>, so no new includes are needed.</p>

<p dir="ltr">I applied this only to <code>qwen3.cpp</code>. The other architecture files build the LM head unconditionally as well, so <strong>this is not Qwen3-specific</strong> — any large-vocab model with a long enough context should reproduce it. A general fix at the <code>llama-graph.cpp</code> level is probably preferable to patching each architecture.</p>

<h2 dir="ltr">Results</h2>

<p dir="ltr">Pristine upstream master, same config as above:</p>

<div dir="ltr">
  | MTL0 compute buffer | 28,001 tokens
-- | -- | --
unpatched | 5283.84 MiB | all-NaN

</div>

<p dir="ltr">Note also that <code>MTL_SHADER_VALIDATION=1</code> reports nothing here. ggml-metal binds one large backend buffer and addresses tensors by offset, so writing into a neighbor tensor stays within the bound buffer's length and is legal from the validator's point of view.</p>

<h2 dir="ltr">Checking whether a build is affected</h2>

<p dir="ltr">Look for <code>MTL0 compute buffer size</code> in the startup log — note this line only appears at <code>-lv 5</code>, the same gating that hides the ggml node logs, so if your launcher does not pass <code>-lv</code> it will not be in your log at all. For <code>Qwen3-Embedding-0.6B</code> at <code>-c 32768 -ub 8192</code>, ~5300 MiB means the LM head is being built; ~475 MiB means it is not.</p>

<p dir="ltr">Separately, and regardless of this issue: an embedding pipeline should validate <code>x is None or x != x</code> per component. A length-only check passes happily when all 1024 components are NaN.</p>

**Comment by ZohnNasse** [comment]
Possibly related: #26044.

That report is CUDA / Volta (sm_70) with Qwen3-Embedding-8B, and describes the same observable failure — all-NaN embedding, server permanently wedged until restart, HTTP 200 throughout, nothing in the log, and correct output under `-ngl 0`. Its isolation table also rules out flash attention, CUDA graphs, `--no-kv-unified`, `--ctx-checkpoints 0`, forcing cuBLAS, quantization level, and toolkit version, which is the same shape of negative results I got on Metal.

I don't want to overclaim a duplicate. Two things don't line up:

- Different backend. None of my node-level evidence transfers to CUDA directly.
- That reporter's trigger is a 698-token chunk, and he concluded it is content-specific rather than length-driven, noting that 212k-token batches embed fine. At vocab 151669, 698 tokens gives an LM head output of 698 × 151669 × 4 ≈ 423 MB — an order of magnitude below the multi-GiB regime I hit.

So: same symptom class, unproven common cause. The cheap thing to check there is whether the LM head is in that graph at all — both models share vocab 151669, and a per-node dump plus the reported compute buffer size would settle it either way. If the LM head is absent from #26044's graph, the two are unrelated and this comment can be disregarded.

**Comment by SecondSeed** [comment]
The issue occurs in the Tensor API implementation of `kernel_mul_mm` in:

```text
ggml/src/ggml-metal/kernels/mul_mm.metal
```

When processing a large number of tokens, `MUL_MAT` can produce a very large F32 output tensor. Its size is:

```text
M * N * batch_count * sizeof(float)
```

As the token dimension increases, this output can exceed 4 GiB. For example, When `M = 151669`, `N = 7343`, and `batch_count = 1` produces 1,113,705,467 float values, requiring 4,454,821,868 bytes, or approximately 4.15 GiB.

The previous implementation represented the entire output batch as one Metal tensor and selected each output tile with `slice(ra, rb)`. Once a tile started beyond the 4 GiB boundary, its global offset could exceed the Tensor API's 32-bit index range, resulting in an incorrect destination address.

I tested this fix on a 16 GB MacBook Air with M5, and the original error no longer occurred. Due to limited hardware resources, I was unable to run the full test suite.

### Fix description
The previous implementation created a tensor view for the entire output batch and then used slice(ra, rb) to select the current output tile:

```metal
device float * dstBatch = (device float *) dst + im * N * M;

auto tD = tensor(
    dstBatch,
    dextents<int32_t, 2>(M, N),
    array<int, 2>({1, M}));

cT.store(tD.slice(ra, rb));
```

The slice coordinates are represented by the tensor API's 32-bit index type. For outputs larger than 4 GiB, the offset of a tile can exceed the supported 32-bit range and produce an incorrect destination address. In addition, `im * N * M` was evaluated using 32-bit integer arithmetic before being added to the pointer.

The new implementation calculates the tile address directly using 64-bit pointer arithmetic:

```metal
const int tileM = min(NRA, M - ra);
const int tileN = min(NRB, N - rb);

device float * dstTile =
    (device float *) dst
    + (uint64_t) im * N * M
    + (uint64_t) rb * M
    + ra;

auto tD = tensor(
    dstTile,
    dextents<int32_t, 2>(tileM, tileN),
    array<int, 2>({1, M}));

cT.store(tD);
```

The large batch and tile offsets are now applied directly to the pointer using `uint64_t`. The tensor API only describes the small local output tile, so it no longer needs to represent the large global slice offset. `tileM` and `tileN` also limit the tensor view at matrix boundaries, ensuring that the final partial tile only writes valid output elements.
